C/STrace/EtwLogger.h

```diff
@@ -0,0 +1,76 @@
+#pragma once
+#include "Interface.h"
+#include "vector.h"
+
+namespace detail
+{
+
+class EtwProviderEvent
+{
+public:
+	EtwProviderEvent();
+	void Destruct();
+
+	NTSTATUS Initialize(const char* eventName, int numberOfFields, va_list fields);
+
+	const char* Name() const;
+	EVENT_DATA_DESCRIPTOR MetadataDescriptor() const;
+
+private:
+	struct EventMetadata
+	{
+		uint16_t TotalLength;
+		uint8_t Tag;
+		char EventName[ANYSIZE_ARRAY];
+	};
+
+	EVENT_DATA_DESCRIPTOR m_eventMetadataDesc;
+};
+
+class EtwProvider
+{
+public:
+	EtwProvider(LPCGUID providerGuid);
+	EtwProvider(const EtwProvider& other) = delete;
+	EtwProvider(EtwProvider&& other);
+	EtwProvider& operator=(const EtwProvider& other) = delete;
+	EtwProvider& operator=(EtwProvider&& other);
+	void Destruct();
+
+	NTSTATUS Initialize(const char* name);
+	NTSTATUS AddEvent(const char* eventName, int numberOfFields, va_list fields);
+	NTSTATUS WriteEvent(const char* eventName, uint8_t eventLevel, uint64_t keyword, int numberOfFields, va_list fields);
+
+	LPCGUID Guid() const noexcept;
```

Comment:
I don't think we need noexcept, either all of these member functions should be noexcept (as we don't use exceptions) or we should omit it entirely. When exceptions are in play there is a difference in optimizations I believe, but since they're off I think there's likely no effect.

Reply:
Force of habit, sorry!